### PR TITLE
SetUp a Rails Api to WanderPal 

### LINKED ISSUE #2: [4pt] Create database and diagram model
- Create the database for backend
- Create the Diagram that displays data relationship and associations (Users/ items )

### LINKED ISSUE #3: [10pt] Generate Models based on diagram
- Generate the models 
- Implement associations and relationship 
- Create migrations. (Create tables) 
- Create validations 
- Create Test for those validations
- Add Cancancan

### LINKED ISSUE #4: [6pt] Generate Api controllers
- Generate Api
- Create methods for create, read, update and delete

### PR DESCRIPTION
**Pull request by Feliverse**
In this PR I fulfilled the following tasks:

Create a new Rails API app called 'WanderPal'.
Initialize the project with Git.
Make sure that the project has a Postgres database set up.
Create endpoints to db items, users, and reservations
Add validations to models
Create unit tests for models
Create an API documentation

[WanderPal frontend ](https://github.com/danielamoreno699/wanderpal-frontend)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,119 @@
...
+
+Please run the following commands `rails db:create` `rails db:migrate` `rails s`
+
+## Tests
+```
+bundle exec rspec
+```
+Or
+```
+rspec spec
```
Comment (Meltrust) [inline README.md:75]: - [x]  You are using Rspec, but the "test" folder still exists, which is used by another test framework. Kindly consider deleting that folder if you are not using it. 👍

Comment (Meltrust) [inline README.md:75]: - [x] When I setup the project  and run the tests according to your instructions, this happens:

image

image

Kindly, make it so your all backend tests pass in order to ensure your app's integrity. 👍

File: app/controllers/api/v1/items_controller.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    else
+      render json: @item.errors, status: :unprocessable_entity
+    end
+  end
+
+  # DELETE /api/v1/items/1
+  def destroy
+    @item.destroy
+    render json: { message: 'Item deleted successfully.' }
+  end
```
Comment (Meltrust) [inline app/controllers/api/v1/items_controller.rb]: - [x] What happens if this action fails?  Kindly, use some error handling in this action. 👍

File: app/controllers/api/v1/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    else
+      render json: @user.errors, status: :unprocessable_entity
+    end
+  end
+
+  # DELETE /api/v1/users/1
+  def destroy
+    @user.destroy
+    render json: { message: 'User deleted successfully.' }
+  end
```
Comment (Meltrust) [inline app/controllers/api/v1/users_controller.rb]: - [x] What happens if this action fails?  Kindly, use some error handling in this action. 👍

**Review by Meltrust — CHANGES_REQUESTED** [review]
Hi @Feliverse,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:

-  Nice code organization ✔️ 
-  API is working well✔️ 
-  Good readme ✔️ 




#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

You can do it

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._







Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------

File: README.md (hunk 1 above)
Comment (Feliverse) [reply README.md:75]: Right, I will delete this files

File: app/controllers/api/v1/items_controller.rb (hunk 2 above)
Comment (Feliverse) [reply app/controllers/api/v1/items_controller.rb]: OK, I will add actions to show fails

File: app/controllers/api/v1/users_controller.rb (hunk 3 above)
Comment (Feliverse) [reply app/controllers/api/v1/users_controller.rb]: OK, I will add actions to show fails

File: app/controllers/api/v1/items_controller.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  # DELETE /api/v1/items/1
+  def destroy
+    def destroy
+      if @item.destroy
+        render json: { message: 'Item deleted successfully.' }
+      else
+        render json: { message: 'Failed to delete item.' }, status: :unprocessable_entity
+      end
+    end
+  end
```
Comment (Meltrust) [inline app/controllers/api/v1/items_controller.rb]: - [ ] You are nesting this method (which means, a method _inside_ another method.

Kindly, unnest the method so it's consistent.  The `destroy` method should look like this:

```ruby
def destroy
    if @item.destroy
      render json: { message: 'Item deleted successfully.' }
    else
      render json: { message: 'Failed to delete item.' }, status: :unprocessable_entity
    end
end
```

File: app/controllers/api/v1/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  # DELETE /api/v1/users/1
+  def destroy
+    def destroy
+      if @user.destroy
+        render json: { message: 'Item deleted successfully.' }
+      else
+        render json: { message: 'Failed to delete item.' }, status: :unprocessable_entity
+      end
+    end
+  end
```
Comment (Meltrust) [inline app/controllers/api/v1/users_controller.rb]: - [ ] You are nesting this method (which means, a method _inside_ another method.

Kindly, unnest the method so it's consistent.  The `destroy` method should look like this:

```ruby
def destroy
  if @user.destroy
    render json: { message: 'Item deleted successfully.' }
  else
    render json: { message: 'Failed to delete item.' }, status: :unprocessable_entity
  end
end
```

**Review by Meltrust — CHANGES_REQUESTED** [review]
Hi @Feliverse,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:

-  Tests are passing✔️
-  Superfluous test folder has been removed✔️ 



#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

You can do it

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._







Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------

File: app/controllers/api/v1/users_controller.rb (hunk 5 above)
Comment (Feliverse) [reply app/controllers/api/v1/users_controller.rb]: ok, I will correct it

**Review by Meltrust — APPROVED** [review]
#### Hi @Feliverse,


# **Wow, you did it 🎉**
Brilliant
## Thank you for the changes implemented 💪 🥇 ㊗️ 
#### _Unless you want to add more features, go ahead to your final presentation ⏩ ⏩ ⏩_ 
## You are about to finish the Microverse program.  You have come a long way!!!




### Good luck in the software industry!!  I'll see you there. ✨
### **Congratulations!!!!!!**  🎉
applause

### To highlight

- Destroy actions now have error handling✔️ 
- Corrected method nesting in controllers✔️ 
- Great job✔️ 



Cheers and Happy coding!👏👏👏


------

**Pull request merged**